### PR TITLE
HTTP bindings

### PR DESCRIPTION
**Pull request by justinboswell**
HTTP works, HTTPS needs work, but it might not be the bindings' fault. Will follow-up after this PR goes in.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws-crt-http/HttpConnection.cs [hunk 1]
```diff
@@ -1,8 +1,245 @@
...
+using System.Security;
+
+using Aws.Crt.IO;
 
 namespace Aws.Crt.Http
 {
-    public class HttpConnection
+    public delegate void OnConnectionSetup(int errorCode);
+    public delegate void OnConnectionShutdown(int errorCode);
+    public delegate void OnStreamOutgoingBody(HttpStream stream, out byte[] buffer);
```
Comment (graebm) [inline aws-crt-http/HttpConnection.cs]: Let's chat about this. Maybe users should just set some kind of stream and we'll read it bit by bit for them. But we could still have an optional callback where they can do so window-adjustment stuff, if they desire.

we might move to an approach like that in C-land too. Brett is doing work on a streaming API for aws-c-io

File: aws-crt-http/HttpConnection.cs [hunk 2]
```diff
@@ -1,8 +1,245 @@
...
+            {
+                API.destroy(handle);
+                return true;
+            }
+        }
+
+        internal Handle NativeHandle { get; private set; }
+
+        public HttpClientConnection Connection { get; protected set; }
+        public int ResponseStatusCode { get; protected set; }
```
Comment (graebm) [inline aws-crt-http/HttpConnection.cs]: ResponseStatusCode is specific to client/request streams
It wouldn't exist on server/response streams

File: aws-crt-http/HttpConnection.cs [hunk 3]
```diff
@@ -1,8 +1,245 @@
...
+
+        public void UpdateWindow(uint incrementSize)
+        {
+            API.update_window(NativeHandle.DangerousGetHandle(), incrementSize);
+        }
+    }
+
+    public sealed class HttpClientStream : HttpStream
+    {
+        public HttpClientStream(HttpClientConnection connection, HttpRequestOptions options)
```
Comment (graebm) [inline aws-crt-http/HttpConnection.cs]: Would it be a better user experience if request creation was a method on the Connection object? I think that's what henso did in aws-crt-cpp. something like:

HttpClientConnection::SendRequest(HttpRequestOptions) -> HttpClientStream

File: aws-crt-http/HttpConnection.cs (hunk 1 above)
Comment (justinboswell) [reply aws-crt-http/HttpConnection.cs]: Discussed offline, we're going to mess with the API for body sending soon.

File: aws-crt-http/HttpConnection.cs (hunk 2 above)
Comment (justinboswell) [reply aws-crt-http/HttpConnection.cs]: Fixed

File: aws-crt-http/HttpConnection.cs (hunk 3 above)
Comment (justinboswell) [reply aws-crt-http/HttpConnection.cs]: I can do both!

File: aws-crt-http/HttpConnection.cs [hunk 4]
```diff
@@ -1,8 +1,245 @@
...
+    internal delegate void OnIncomingHeaderBlockDoneNative(bool hasBody);
+    internal delegate void OnIncomingBodyNative(
+        [In, MarshalAs(UnmanagedType.LPArray, SizeParamIndex=1)] byte[] buffer, 
+        UInt64 size);
+    internal delegate void OnStreamCompleteNative(int errorCode);
+
+    public sealed class HttpRequestOptions
+    {
+        public string Method { get; set; }
+        public Uri Uri { get; set; }
```
[the remark this answers is not in the document]
Comment (justinboswell) [reply aws-crt-http/HttpConnection.cs]: Done did it

File: native/src/http_client.c [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+    const char *host_name,
+    uint16_t port,
+    struct aws_socket_options *socket_options,
+    struct aws_tls_connection_options *tls_connection_options,
+    aws_dotnet_http_on_client_connection_setup_fn *on_setup,
+    aws_dotnet_http_on_client_connection_shutdown_fn *on_shutdown) {
+
+    struct aws_allocator *allocator = aws_dotnet_get_allocator();
+    struct aws_dotnet_http_connection *connection =
+        aws_mem_acquire(allocator, sizeof(struct aws_dotnet_http_connection));
```
Comment (graebm) [inline native/src/http_client.c]: not checking result

**Pull request merged**